File: service/src/main/java/by/rekuts/travelagency/service/impl/UserServiceImpl.java
package by.rekuts.travelagency.service.impl;

import by.rekuts.travelagency.domain.Tour;
import by.rekuts.travelagency.domain.User;
import by.rekuts.travelagency.repository.UserRepository;
import by.rekuts.travelagency.repository.UserSpecification;
import by.rekuts.travelagency.service.UserService;
import lombok.RequiredArgsConstructor;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Collections;
import java.util.List;
import java.util.Optional;

@RequiredArgsConstructor
@Service
public class UserServiceImpl implements UserService, UserDetailsService {

    private final UserRepository userRepository;

    @Transactional
    @Override
    public void insert(User user) {
        userRepository.save(user);
    }

    @Transactional
    @Override
    public void update(User user) {
        userRepository.save(user);
    }

    @Transactional
    @Override
    public void delete(int id) {
        userRepository.deleteById(id);
    }

    @Transactional
    @Override
    public List<User> getList(Specification<User> specification) {
        return userRepository.findAll(specification);
    }

    @Transactional(readOnly = true)
    @Override
    public UserDetails loadUserByUsername(String login) {
        UserSpecification specification = new UserSpecification();
        specification.setLogin(login);
        User user = userRepository.findAll(specification).get(0);
        if (user != null) {
            return new org.springframework.security.core.userdetails.User(
                    user.getLogin(),
                    user.getPassword(),
                    Collections.singletonList(new SimpleGrantedAuthority(user.getRole().name()))
            );
        }
        throw new UsernameNotFoundException("User '" + login + "' not found");
    }

    @Override
    public void addTourToFavorites(User user, Tour tour) {
        Optional<User> temp = userRepository.findById(user.getId());
        temp.ifPresent(u -> {
            u.getTours().add(tour);
            userRepository.save(u);
        });
    }

    @Override
    public void removeTourFromFavorites(User user, Tour tour) {
        Optional<User> temp = userRepository.findById(user.getId());
        temp.ifPresent(u -> {
            u.getTours().remove(tour);
            userRepository.save(u);
        });
    }
}

File: service/src/main/scala/by/rekuts/travelagency/service/impl/UserServiceImpl.scala
package by.rekuts.travelagency.service.impl

import java.util
import java.util.Collections

import by.rekuts.travelagency.domain.{Tour, User}
import by.rekuts.travelagency.repository.{Specification, UserRepository, UserSpecification}
import by.rekuts.travelagency.service.UserService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.{UserDetails, UserDetailsService, UsernameNotFoundException}
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class UserServiceImpl extends UserService with UserDetailsService {
  @Autowired
  private var userRepository: UserRepository = _

  @Transactional
  override def insert(user: User): Unit = userRepository.insert(user)

  @Transactional
  override def update(user: User): Unit = userRepository.update(user)

  @Transactional
  override def delete(id: Int): Unit = userRepository.delete(id)

  @Transactional
  override def getList(specification: Specification[_]): util.List[User] = userRepository.getList(specification)

  @Transactional(readOnly = true)
  override def loadUserByUsername(login: String): UserDetails = {
    val specification: UserSpecification = new UserSpecification
    specification.login_$eq(login)
    val user: User = userRepository.getList(specification).get(0)
    if (user != null) return new org.springframework.security.core.userdetails.User(
      user.login,
      user.password,
      Collections.singletonList(new SimpleGrantedAuthority(user.role.toString))
    )
    throw new UsernameNotFoundException("User '" + login + "' not found")
  }

  override def addTourToFavorites(user: User, tour: Tour): Unit = userRepository.addTourToFavorites(user, tour)

  override def removeTourFromFavorites(user: User, tour: Tour): Unit = userRepository.removeTourFromFavorites(user, tour)
}
